src/vm/amsi.cpp

```diff
@@ -0,0 +1,147 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+//
+// File: amsi.cpp
+//
+
+#include "common.h"
+#include "amsi.h"
+
+namespace
+{
+    // https://docs.microsoft.com/en-us/windows/desktop/api/amsi/
+    DECLARE_HANDLE(HAMSICONTEXT);
+    DECLARE_HANDLE(HAMSISESSION);
+
+    enum AMSI_RESULT
+    {
+        AMSI_RESULT_CLEAN                   = 0,
+        AMSI_RESULT_NOT_DETECTED            = 1,
+        AMSI_RESULT_BLOCKED_BY_ADMIN_START  = 0x4000,
+        AMSI_RESULT_BLOCKED_BY_ADMIN_END    = 0x4fff,
+        AMSI_RESULT_DETECTED                = 0x8000
+    }   AMSI_RESULT;
+
+    bool AmsiResultIsMalware(DWORD result)
+    {
+        return result >= AMSI_RESULT_DETECTED;
+    }
+
+    bool AmsiResultIsBlockedByAdmin(DWORD result)
+    {
+        return result >= AMSI_RESULT_BLOCKED_BY_ADMIN_START
+            && result <= AMSI_RESULT_BLOCKED_BY_ADMIN_END;
+    }
+
+    using PAMSI_AMSISCANBUFFER_API = HRESULT(WINAPI *)(
+        _In_ HAMSICONTEXT amsiContext,
+        _In_ PVOID buffer,
+        _In_ ULONG length,
+        _In_ LPCWSTR contentName,
+        _In_opt_ HAMSISESSION session,
+        _Out_ DWORD *result);
+
+    using PAMSI_AMSIINITIALIZE_API = HRESULT(WINAPI *)(
+        _In_ LPCWSTR appName,
+        _Out_ HAMSICONTEXT *amsiContext);
+
+    PAMSI_AMSISCANBUFFER_API AmsiScanBuffer;
+    HAMSICONTEXT s_amsiContext;
+    CRITSEC_COOKIE s_csAmsi;
+
+    bool InitializeLock()
+    {
+        if (s_csAmsi != nullptr)
+            return true;
+
+        CRITSEC_COOKIE lock = ClrCreateCriticalSection(CrstLeafLock, CRST_REENTRANCY);
+        if (lock == nullptr)
+            return false;
+
+        if (InterlockedCompareExchangeT<CRITSEC_COOKIE>(&s_csAmsi, lock, nullptr) != nullptr)
+            ClrDeleteCriticalSection(lock);
+
+        return true;
+    }
+
+    bool HasLoadLibrarySearchSystem32()
```

Comment:
There's no need to test for OS support for KB2533623 or LOAD_LIBRARY_SEARCH_SYSTEM32. The patch has been out for 8½ years. There's no in-market operating system under active support that doesn't already include the patch as part of its base image.

Reply:
Ah, thanks. Removed.